{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "6862f8c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Building DAG of jobs...\n",
      "Using shell: /lustre1/project/stg_00079/teaching/I0U19a_conda_2024/bin/bash\n",
      "Provided cores: 1 (use --cores to define parallelism)\n",
      "Rules claiming more threads will be scaled down.\n",
      "Job stats:\n",
      "job              count\n",
      "-------------  -------\n",
      "SNP_calling          1\n",
      "SNP_norm_filt        1\n",
      "all                  1\n",
      "bwa                  1\n",
      "fastqc               1\n",
      "snpeff               1\n",
      "total                6\n",
      "\n",
      "Select jobs to execute...\n",
      "\n",
      "[Wed May  1 12:15:55 2024]\n",
      "rule bwa:\n",
      "    input: /staging/leuven/stg_00079/teaching/1000genomes/HG00141.GRCh38DH.exome.chr21.fq.gz\n",
      "    output: 2_bwa/HG00141.GRCh38DH.exome.chr21.bam, 2_bwa/HG00141.GRCh38DH.exome.chr21.bam.bai\n",
      "    jobid: 5\n",
      "    reason: Missing output files: 2_bwa/HG00141.GRCh38DH.exome.chr21.bam\n",
      "    wildcards: sample=HG00141.GRCh38DH.exome.chr21\n",
      "    resources: tmpdir=/tmp\n",
      "\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[M::process] read 99010 sequences (10000010 bp)...\n",
      "[M::process] read 99010 sequences (10000010 bp)...\n",
      "[M::mem_process_seqs] Processed 99010 reads in 4.464 CPU sec, 4.347 real sec\n",
      "[M::process] read 99010 sequences (10000010 bp)...\n",
      "[M::mem_process_seqs] Processed 99010 reads in 3.931 CPU sec, 3.790 real sec\n",
      "[M::process] read 99010 sequences (10000010 bp)...\n",
      "[M::mem_process_seqs] Processed 99010 reads in 3.645 CPU sec, 3.502 real sec\n",
      "[M::process] read 99010 sequences (10000010 bp)...\n",
      "[M::mem_process_seqs] Processed 99010 reads in 3.542 CPU sec, 3.389 real sec\n",
      "[M::process] read 99010 sequences (10000010 bp)...\n",
      "[M::mem_process_seqs] Processed 99010 reads in 3.552 CPU sec, 3.423 real sec\n",
      "[M::process] read 67284 sequences (6795684 bp)...\n",
      "[M::mem_process_seqs] Processed 99010 reads in 4.433 CPU sec, 4.339 real sec\n",
      "[M::mem_process_seqs] Processed 67284 reads in 2.193 CPU sec, 2.153 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem /lustre1/project/stg_00079/teaching/hg38_21/chr21.fa /staging/leuven/stg_00079/teaching/1000genomes/HG00141.GRCh38DH.exome.chr21.fq.gz\n",
      "[main] Real time: 25.178 sec; CPU: 25.986 sec\n",
      "[Wed May  1 12:16:26 2024]\n",
      "Finished job 5.\n",
      "1 of 6 steps (17%) done\n",
      "Select jobs to execute...\n",
      "\n",
      "[Wed May  1 12:16:26 2024]\n",
      "rule SNP_calling:\n",
      "    input: /lustre1/project/stg_00079/teaching/hg38_21/chr21.fa, 2_bwa/HG00141.GRCh38DH.exome.chr21.bam\n",
      "    output: 3_samtools/snps.vcf\n",
      "    jobid: 4\n",
      "    reason: Missing output files: 3_samtools/snps.vcf; Input files updated by another job: 2_bwa/HG00141.GRCh38DH.exome.chr21.bam\n",
      "    resources: tmpdir=/tmp\n",
      "\n",
      "Note: none of --samples-file, --ploidy or --ploidy-file given, assuming all sites are diploid\n",
      "[mpileup] 1 samples in 1 input files\n",
      "[mpileup] maximum number of reads per input file set to -d 250\n",
      "[Wed May  1 12:16:40 2024]\n",
      "Finished job 4.\n",
      "2 of 6 steps (33%) done\n",
      "Select jobs to execute...\n",
      "\n",
      "[Wed May  1 12:16:40 2024]\n",
      "rule SNP_norm_filt:\n",
      "    input: /lustre1/project/stg_00079/teaching/hg38_21/chr21.fa, 3_samtools/snps.vcf\n",
      "    output: 4_cleaned/snps.cleaned.vcf\n",
      "    jobid: 3\n",
      "    reason: Missing output files: 4_cleaned/snps.cleaned.vcf; Input files updated by another job: 3_samtools/snps.vcf\n",
      "    resources: tmpdir=/tmp\n",
      "\n",
      "decompose v0.5\n",
      "\n",
      "options:     input VCF file        -\n",
      "         [s] smart decomposition   false (experimental)\n",
      "         [o] output VCF file       -\n",
      "\n",
      "uniq v0.57normalize v\n",
      "\n",
      "0.5options:     input VCF file        \n",
      "-\n",
      "\n",
      "options:     input VCF file                                           [o] output VCF file       -\n",
      "-         [o] output VCF file                                 \n",
      "-\n",
      "         [w] sorting window size                             10000\n",
      "         [n] no fail on reference inconsistency for non SNPs true\n",
      "         [q] quiet                                           false\n",
      "         [d] debug                                           false\n",
      "         [r] reference FASTA file                            /lustre1/project/stg_00079/teaching/hg38_21/chr21.fa\n",
      "\n",
      "\n",
      "stats: no. variants                 : 8061\n",
      "       no. biallelic variants       : 8052\n",
      "       no. multiallelic variants    : 9\n",
      "\n",
      "       no. additional biallelics    : 9\n",
      "       total no. of biallelics      : 8070\n",
      "\n",
      "Time elapsed: 0.03s\n",
      "\n",
      "\n",
      "stats: biallelic\n",
      "          no. left trimmed                      : 0\n",
      "          no. right trimmed                     : 329\n",
      "          no. left and right trimmed            : 0\n",
      "          no. right trimmed and left aligned    : 0\n",
      "          no. left aligned                      : 0\n",
      "\n",
      "       total no. biallelic normalized           : 329\n",
      "\n",
      "       multiallelic\n",
      "          no. left trimmed                      : 0\n",
      "          no. right trimmed                     : 0\n",
      "          no. left and right trimmed            : 0\n",
      "          no. right trimmed and left aligned    : 0\n",
      "          no. left aligned                      : 0\n",
      "\n",
      "       total no. multiallelic normalized        : 0\n",
      "\n",
      "       total no. variants normalized            : 329\n",
      "       total no. variants observed              : 8070\n",
      "       total no. reference observed             : 0\n",
      "\n",
      "Time elapsed: 0.04s\n",
      "\n",
      "\n",
      "stats: Total number of observed variants   8070\n",
      "       Total number of unique variants     8070\n",
      "\n",
      "Time elapsed: 0.03s\n",
      "\n",
      "[Wed May  1 12:16:40 2024]\n",
      "Finished job 3.\n",
      "3 of 6 steps (50%) done\n",
      "Select jobs to execute...\n",
      "\n",
      "[Wed May  1 12:16:40 2024]\n",
      "rule fastqc:\n",
      "    input: /staging/leuven/stg_00079/teaching/1000genomes/HG00141.GRCh38DH.exome.chr21.fq.gz\n",
      "    output: 1_fastqc/HG00141.GRCh38DH.exome.chr21_fastqc.zip, 1_fastqc/HG00141.GRCh38DH.exome.chr21_fastqc.html, 1_fastqc/HG00141.GRCh38DH.exome.chr21_fastqc/summary.txt, 1_fastqc/HG00141.GRCh38DH.exome.chr21_fastqc/Images/per_base_quality.png\n",
      "    jobid: 1\n",
      "    reason: Missing output files: 1_fastqc/HG00141.GRCh38DH.exome.chr21_fastqc.zip\n",
      "    wildcards: name=HG00141.GRCh38DH.exome.chr21\n",
      "    resources: tmpdir=/tmp\n",
      "\n",
      "Input Fastq: /staging/leuven/stg_00079/teaching/1000genomes/HG00141.GRCh38DH.exome.chr21.fq.gz \n",
      "application/gzip\n",
      "Started analysis of HG00141.GRCh38DH.exome.chr21.fq.gz\n",
      "Approx 5% complete for HG00141.GRCh38DH.exome.chr21.fq.gz\n",
      "Approx 10% complete for HG00141.GRCh38DH.exome.chr21.fq.gz\n",
      "Approx 15% complete for HG00141.GRCh38DH.exome.chr21.fq.gz\n",
      "Approx 20% complete for HG00141.GRCh38DH.exome.chr21.fq.gz\n",
      "Approx 25% complete for HG00141.GRCh38DH.exome.chr21.fq.gz\n",
      "Approx 30% complete for HG00141.GRCh38DH.exome.chr21.fq.gz\n",
      "Approx 35% complete for HG00141.GRCh38DH.exome.chr21.fq.gz\n",
      "Approx 40% complete for HG00141.GRCh38DH.exome.chr21.fq.gz\n",
      "Approx 45% complete for HG00141.GRCh38DH.exome.chr21.fq.gz\n",
      "Approx 50% complete for HG00141.GRCh38DH.exome.chr21.fq.gz\n",
      "Approx 55% complete for HG00141.GRCh38DH.exome.chr21.fq.gz\n",
      "Approx 60% complete for HG00141.GRCh38DH.exome.chr21.fq.gz\n",
      "Approx 65% complete for HG00141.GRCh38DH.exome.chr21.fq.gz\n",
      "Approx 70% complete for HG00141.GRCh38DH.exome.chr21.fq.gz\n",
      "Approx 75% complete for HG00141.GRCh38DH.exome.chr21.fq.gz\n",
      "Approx 80% complete for HG00141.GRCh38DH.exome.chr21.fq.gz\n",
      "Approx 85% complete for HG00141.GRCh38DH.exome.chr21.fq.gz\n",
      "Approx 90% complete for HG00141.GRCh38DH.exome.chr21.fq.gz\n",
      "Approx 95% complete for HG00141.GRCh38DH.exome.chr21.fq.gz\n",
      "Analysis complete for HG00141.GRCh38DH.exome.chr21.fq.gz\n",
      "Unzip the output\n",
      "Archive:  HG00141.GRCh38DH.exome.chr21_fastqc.zip\n",
      "   creating: HG00141.GRCh38DH.exome.chr21_fastqc/Icons/\n",
      "  inflating: HG00141.GRCh38DH.exome.chr21_fastqc/Icons/fastqc_icon.png  \n",
      "  inflating: HG00141.GRCh38DH.exome.chr21_fastqc/Icons/warning.png  \n",
      "  inflating: HG00141.GRCh38DH.exome.chr21_fastqc/Icons/error.png  \n",
      "  inflating: HG00141.GRCh38DH.exome.chr21_fastqc/Icons/tick.png  \n",
      "  inflating: HG00141.GRCh38DH.exome.chr21_fastqc/summary.txt  \n",
      "  inflating: HG00141.GRCh38DH.exome.chr21_fastqc/Images/per_base_quality.svg  \n",
      "  inflating: HG00141.GRCh38DH.exome.chr21_fastqc/Images/per_base_quality.png  \n",
      "  inflating: HG00141.GRCh38DH.exome.chr21_fastqc/Images/per_sequence_quality.svg  \n",
      "  inflating: HG00141.GRCh38DH.exome.chr21_fastqc/Images/per_sequence_quality.png  \n",
      "  inflating: HG00141.GRCh38DH.exome.chr21_fastqc/Images/per_base_sequence_content.svg  \n",
      "  inflating: HG00141.GRCh38DH.exome.chr21_fastqc/Images/per_base_sequence_content.png  \n",
      "  inflating: HG00141.GRCh38DH.exome.chr21_fastqc/Images/per_sequence_gc_content.svg  \n",
      "  inflating: HG00141.GRCh38DH.exome.chr21_fastqc/Images/per_sequence_gc_content.png  \n",
      "  inflating: HG00141.GRCh38DH.exome.chr21_fastqc/Images/per_base_n_content.svg  \n",
      "  inflating: HG00141.GRCh38DH.exome.chr21_fastqc/Images/per_base_n_content.png  \n",
      "  inflating: HG00141.GRCh38DH.exome.chr21_fastqc/Images/sequence_length_distribution.svg  \n",
      "  inflating: HG00141.GRCh38DH.exome.chr21_fastqc/Images/sequence_length_distribution.png  \n",
      "  inflating: HG00141.GRCh38DH.exome.chr21_fastqc/Images/duplication_levels.svg  \n",
      "  inflating: HG00141.GRCh38DH.exome.chr21_fastqc/Images/duplication_levels.png  \n",
      "  inflating: HG00141.GRCh38DH.exome.chr21_fastqc/Images/adapter_content.svg  \n",
      "  inflating: HG00141.GRCh38DH.exome.chr21_fastqc/Images/adapter_content.png  \n",
      "  inflating: HG00141.GRCh38DH.exome.chr21_fastqc/fastqc_report.html  \n",
      "  inflating: HG00141.GRCh38DH.exome.chr21_fastqc/fastqc_data.txt  \n",
      "  inflating: HG00141.GRCh38DH.exome.chr21_fastqc/fastqc.fo  \n",
      "[Wed May  1 12:16:45 2024]\n",
      "Finished job 1.\n",
      "4 of 6 steps (67%) done\n",
      "Select jobs to execute...\n",
      "\n",
      "[Wed May  1 12:16:45 2024]\n",
      "rule snpeff:\n",
      "    input: 4_cleaned/snps.cleaned.vcf\n",
      "    output: 5_snpeff/snps.annotated.vcf, 5_snpeff/snpEff_summary.html, 5_snpeff/snpEff_genes.txt\n",
      "    log: 5_snpeff/snakemake.err\n",
      "    jobid: 2\n",
      "    reason: Missing output files: 5_snpeff/snps.annotated.vcf; Input files updated by another job: 4_cleaned/snps.cleaned.vcf\n",
      "    resources: tmpdir=/tmp\n",
      "\n",
      "Error: Unable to access jarfile /lustre1/project/stg_00079/teaching/miniconda3/pkgs/snpeff-5.1-hdfd78af_2/share/snpeff-5.1-2/snpEff.jar\n",
      "[Wed May  1 12:16:45 2024]\n",
      "Error in rule snpeff:\n",
      "    jobid: 2\n",
      "    input: 4_cleaned/snps.cleaned.vcf\n",
      "    output: 5_snpeff/snps.annotated.vcf, 5_snpeff/snpEff_summary.html, 5_snpeff/snpEff_genes.txt\n",
      "    log: 5_snpeff/snakemake.err (check log file(s) for error details)\n",
      "    shell:\n",
      "        \n",
      "\n",
      "        mkdir -p 5_snpeff\n",
      "\n",
      "        java -Xmx4096m -jar             /lustre1/project/stg_00079/teaching/miniconda3/pkgs/snpeff-5.1-hdfd78af_2/share/snpeff-5.1-2/snpEff.jar eff hg19             -dataDir /staging/leuven/stg_00079/teaching/snpeff_db/             4_cleaned/snps.cleaned.vcf > 5_snpeff/snps.annotated.vcf\n",
      "\n",
      "        # move output files to the snpeff output folder\n",
      "        mv snpEff_genes.txt snpEff_summary.html 5_snpeff\n",
      "\n",
      "        \n",
      "        (one of the commands exited with non-zero exit code; note that snakemake uses bash strict mode!)\n",
      "\n",
      "Removing output files of failed job snpeff since they might be corrupted:\n",
      "5_snpeff/snps.annotated.vcf\n",
      "Shutting down, this might take some time.\n",
      "Exiting because a job execution failed. Look above for error message\n",
      "Complete log: .snakemake/log/2024-05-01T121554.882807.snakemake.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Define the command to run Snakemake\n",
    "!snakemake -c1 -d /data/leuven/362/vsc36251/Assignment2/snakemakeOutput -s /data/leuven/362/vsc36251/Assignment2/snakefile > snakemake_output.log 2>&1\n",
    "\n",
    "with open(\"snakemake_output.log\", \"r\") as file:\n",
    "    log_contents = file.read()\n",
    "\n",
    "# Print the contents of the file\n",
    "print(log_contents)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "large_omics_2023",
   "language": "python",
   "name": "large_omics_2023"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
